wait until actually stopped to tell windows we are

tewinget opened the pull request:
We should send STOP_PENDING rather than STOPPED while we aren't yet actually stopped; STOPPED is already sent when we actually finish stopping.

Also fixes some silly argc/argv shenanigans, I think.

TODO:
- [x] delay windows service notification until after we are fully initialized so that if we fail (e.g. can't start wintun, config error, etc.) we won't have told windows that we started (and it can properly report a failure to start the service).

tewinget committed:
wait until actually stopped to tell windows we are

We should send STOP_PENDING rather than STOPPED while we aren't yet
actually stopped; STOPPED is already sent when we actually finish
stopping.

Also fixes some silly argc/argv shenanigans, I think.

tewinget committed:
debian missing yacc apparently all of a sudden

majestrate commented on daemon/lokinet.cpp:
```diff
@@ -681,10 +681,12 @@ win32_daemon_entry(DWORD argc, LPTSTR* argv)
+  // SCM calls this function with different args than a normal "main" expects,
+  // but lokinet_main expects normal args, so set them here instead.  At the
+  // moment we are not passing any args to lokinet_main this way anyway though.
+  argc = 1;
```
i would suggest using a totally separate array here so that we arent screwing with argc and argv

majestrate committed:
correct windows service manager behavior.

report status to window service manager when we get and iterogate
message from the service manager.
update comments to reflect these changes.

tewinget committed:
some useful log statements

majestrate committed:
fix comment to reflect reality

majestrate committed:
system layer manager

add an interface type for platform dependant system layer behavior.
this gives lokinet a platform agnostic way to inform the system layer
of its current state.

majestrate committed:
fix ifdef guards

majestrate committed:
turn off service manager in unit tests

jagerman committed:
Remove dead/redundant code

- win32_platform.cpp is dead
- win32_platform.hpp is useless

Style changes from clang-tidy warnings:
- remove `virtual` from some definitions that already have `override`
- remove virtual destructor from NetworkInterface because it already has
  a virtual destructor via the base type (and clang-tiny warns about it)

jagerman committed:
Fix crashes in wintun and windivert stopping

Fixes windows shutdown crashes:

- windivert wasn't handling an ERROR_NO_DATA, which it gets when
  finished handling everything after a shutdown.
- wintun ReadPacket still gets invoked after end_session is called, but
  shouldn't be.  This adds an atomic<bool> to early return.
- fixes up some settings we send for windows service manager notify

jagerman committed:
Add some more debugging

jagerman committed:
🤷 Work around ub_ctx_delete crash

When unbound is in threaded async mode on Windows it crashes in
ub_ctx_delete during shutdown for unknown reasons, but only when stopped
via windows service control.

Switch it to process async mode which doesn't crash.

jagerman committed:
Update gui to latest

jagerman committed:
Add networkReady to stats result

majestrate reviewed (approved):
if it works now it looks fine

jagerman commented on llarp/linux/service_manager.cpp:
```diff
@@ -0,0 +1,61 @@
+      if (m_State == ServiceState::Starting)
+      {
+        // TODO: maybe request more time?
+        ::sd_notify(0, "READY=0");
```
This isn't right; from sd_notify docs:

> Since there is little value in signaling non-readiness, the only value services should send is "READY=1" (i.e. "READY=0" is not defined).

jagerman commented on llarp/linux/service_manager.cpp:
```diff
@@ -0,0 +1,61 @@
+#if defined(WITH_SYSTEMD)
+#include <systemd/sd-daemon.h>
+#include <cassert>
+#endif
```
I feel like this should get split into two separate files: A no-op version, and a systemd version and we conditionally compile base on whether we have systemd support.

jagerman committed:
Extend windows startup timeout

If wintun fails it seems to take about 15s, so extend the startup
timeout so that it can fail gracefully (and let us clean up before
exiting).

Also refactors the timeouts to chrono constants.

majestrate committed:
fixups

* move nop service manager to its own compilation unit and remove old
compilation units (touches cmake)
* change service manager interface to also be able to report stats
* fixup service manager interface's method names
* move sd_notify to service manager

majestrate committed:
format

jagerman commented on llarp/linux/service_manager.cpp:
```diff
@@ -1,13 +1,10 @@
 #include <llarp/util/service_manager.hpp>
```
I think this file should be removed now?

majestrate replied:
yes it should be removed

jagerman commented on llarp/router/router.cpp:
```diff
@@ -874,6 +874,60 @@ namespace llarp
+  Router::status_line()
+  {
+    std::string status;
+#if defined(WITH_SYSTEMD)
```
Should nuke this ifdef guard; this status line doesn't depend on systemd anymore.

majestrate replied:
done

jagerman commented on llarp/router/router.cpp:
```diff
@@ -874,6 +874,60 @@ namespace llarp
+    std::string status;
+#if defined(WITH_SYSTEMD)
+    auto out = std::back_inserter(status);
+    fmt::format_to(out, "WATCHDOG=1\nSTATUS=v{}", llarp::VERSION_STR);
```
The return value here should be starting from "v", but not including the `WATCHDOG=1\nSTATUS=` prefix (which should be the responsibility of sd_service_manager to add).

majestrate replied:
done

majestrate committed:
clean up service manager for systemd

* remove old compilation unit
* move systemd specific parts of status to service manager
* make router status not systemd specific

jagerman commented on llarp/router/router.cpp:
```diff
@@ -870,6 +874,58 @@ namespace llarp
+  {
+    std::string status;
+    auto out = std::back_inserter(status);
+    fmt::format_to(out, "v{}", llarp::VERSION_STR);
```
This appears to be the only place that uses `VERSION_STR` in the entire code base; should we just drop it entirely and use fmt here to join the numeric array like this?
```suggestion
    fmt::format_to(out, "v{}", fmt::join(llarp::VERSION, "."));
```

majestrate committed:
Update llarp/router/router.cpp

use `fmt::join` for lokinet version string in status

Co-authored-by: Jason Rhinelander <jason@imaginary.ca>

majestrate committed:
simplify logic for disabling service manager on windows

majestrate committed:
fix typo in filename

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -84,9 +86,6 @@ install_win32_daemon()
   }
-  // just put the flag here. we eat it later on and specify the
-  // config path in the daemon entry point
-  StringCchCat(szPath.data(), 1024, " --win32-daemon");
```
This seems wrong.  If removed here it seems there is nowhere that sets the service up with the `--win32-daemon` flag, but main() still relies on it?

majestrate replied:
yea. i also dont know.

jagerman replied:
Why did you remove it?

majestrate replied:
it was using a function called `StringCchCat` from some windows header that does god knows what. i forget to figure out wtf it did and emulate the desired behavior.

jagerman commented on daemon/lokinet.cpp:
```diff
@@ -82,9 +84,6 @@ install_win32_daemon()
   }
-  // just put the flag here. we eat it later on and specify the
-  // config path in the daemon entry point
-  StringCchCat(szPath.data(), 1024, " --win32-daemon");
```
This seems wrong.  If removed here it seems there is nowhere that sets the service up with the `--win32-daemon` flag, but main() still relies on it?

jagerman committed:
Move log init even earlier

jagerman committed:
Remove unused VERSION_STR constant

jagerman committed:
Improve windows running-as-a-service detection works

Get rid of the --win32-daemon hack (which was removed from the service
itself earlier in this PR, by mistake) and replace it with detection of
the error code for "not running as a service" that windows gives us back
if we try to set up service controller dispatching but aren't a service.

tewinget committed:
remove dead dns resolver code

majestrate committed:
fix crash on shutdown

we were calling llarp::Context::HandleSignal from a non mainloop
thread when running as a win32 service. this caused issues with a non
clean destruction.

call our signal handler instead of llarp::Context::HandleSignal

jagerman committed:
Revert "🤷 Work around ub_ctx_delete crash"

This reverts commit 3f2311887ffb4082d4891da7332b26c6dae4ce6a.

jagerman committed:
Don't raise log level on shutdown

If already below info (e.g. debug) it should stay there; we only want to
*lower* it to info if above info.

jagerman committed:
Remove bad assert

We do and should be able to call this multiple times during shutdown to
signal that we are advancing through shutdown.

jagerman committed:
Add more stopping signals

tewinget committed:
stop-time debug statements

jagerman committed:
windivert: avoid trying to send during shutdown

Occasionally during shutdown windivert will crash because a thread tries
sending after we've called wd::shutdown, which isn't allowed.  Add an
atomic bool to prevent this.

majestrate committed:
rewordme

majestrate committed:
Merge remote-tracking branch 'tom-ssh/windows-service-issues' into windows-service-issues

majestrate committed:
format pedant

majestrate committed:
dont clear m_pending on shutdown that is taken care of that for us on win32 make sure we do not query when we are shutting down

majestrate committed:
i hate this, a query's packetsource_base needed to be a shared_ptr because the query is kept alive by it (sometimes)

majestrate committed:
fix previous commit

majestrate committed:
fix typo remove comment

majestrate committed:
remove pending query after sending a reply of any kind

jagerman committed:
More debug

majestrate committed:
use weak_from_this()

majestrate committed:
format

jagerman committed:
Fix crashy race condition in shutdown

Query->Cancel() will remove the Query, but that introduces a race
condition where unbound may still try to invoke the callback (with a
no-longer-valid pointer) if we do it before the ub_ctx_delete call.

Move to it afterwards so that we only cancel things that unbound didn't

jagerman committed:
Fix crash on unbound cleanup

We need to make a copy here because (see comment).

jagerman committed:
patch unbound to fix windows shutdown crash

jagerman committed:
RPC: fix "halt" command

tewinget commented:
supplanted by #2045
